test/xbgpu/test_engine.py

```diff
@@ -565,11 +540,23 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
             )
 
         with PromDiff(namespace=METRIC_NAMESPACE) as prom_diff:
-            # Generate one extra chunk to simulate an incomplete accumulation
-            # to check that dumps are aligned correctly - even if the first
-            # received batch is from the middle of an accumulation.
-            batch_start_index = first_batch_index * max_heap_acc_threshold - HEAPS_PER_FENGINE_PER_CHUNK
-            batch_end_index = (first_batch_index + min_full_accumulations) * max_heap_acc_threshold
+            # NOTE: The product of `heap_acc_thresholds` is used in two ways
+            # below. Both uses are to ensure there is a whole number of
+            # accumulations for *both* XPipelines. The first usage is
+            # dual-purpose once more:
+            # - In addition to the above, the arbitrary start position for data
+            #   in this test dictates the first received batch to be in the
+            #   middle of an accumulation.
+            # - More accurately, it forces the first accumulation (to be
+            #   processed) to only contain one batch of data. This ensures
+            #   we test that output dumps are aligned correctly, despite
+            #   the first data processed not being on an accumulation
+            #   boundary.
+            heap_acc_thresholds = [corrprod_output.heap_accumulation_threshold for corrprod_output in corrprod_outputs]
+            first_batch_index = 12 * np.product(heap_acc_thresholds)
+            batch_start_index = first_batch_index - HEAPS_PER_FENGINE_PER_CHUNK
```

Comment:
This doesn't match the comment, which says "one batch of data". Will the test even do the right thing if HEAPS_PER_FENGINE_PER_CHUNK is >= one of the heap accumulation thresholds?

Reply:
Ah, yes, well, this was my original observation and I didn't really know whether to stick or twist with 
* adding logic to prevent that from happening, and/or
* making a note that we've chosen values ~carefully to ensure this does not happen.

Any recommendations/suggestions/requests? I could probably do something at the top of the test like
* `assert heap_acc_threshold > HEAPS_PER_FENGINE_PER_CHUNK` for each threshold?

Reply:
Explicitly requested a re-review from you mostly to catch your attention to my reply above.

Reply:
I'd be happy with an assert statement such as you propose, @amishatishpatel, perhaps accompanied by a comment explaining that we haven't verified the logic of the test for if that constraint isn't met.

Reply:
What is the reason for using HEAPS_PER_FENGINE_PER_CHUNK here instead of 1?

Reply:
> What is the reason for using HEAPS_PER_FENGINE_PER_CHUNK [2] here instead of 1?

My guess is that it's to be explicit that an entire Chunk went missing.
- Previously the `heap_accumulation_threshold` was set to 4,
- So the comment of "testing `if the first received batch is from the **middle** of an accumulation`" held true because 2 heaps (1 Chunk) were halfway to a total of 4 heaps (2 Chunks) required for an accumulation.

I figured I'd do a little `git blame` to see where that came from, and wouldn't you know [the commit](https://github.com/ska-sa/katgpucbf/commit/db1bae921e4b3b2be9ac3cd9e2be661e543c4139) was made by one @bmerry. I gather your reasoning was similar?

Either way, I think I'm going to do the assert statement and comment to be more explicit about this bit of code.

Reply:
Hoisted by my own petard!

> Either way, I think I'm going to do the assert statement and comment to be more explicit about this bit of code.

Sounds sensible.